internal/oci/runtime_oci.go

```diff
@@ -616,33 +616,68 @@ func waitContainerStop(ctx context.Context, c *Container, timeout time.Duration,
 			}
 		}
 	}()
-	select {
-	case <-done:
-		return nil
-	case <-ctx.Done():
-		close(chControl)
-		return ctx.Err()
-	case <-time.After(timeout):
-		close(chControl)
-		if ignoreKill {
-			return fmt.Errorf("timeout reached after %.0f seconds waiting for container process to exit",
-				timeout.Seconds())
-		}
-		pid, err := c.pid()
-		if err != nil {
-			return err
-		}
-		if err := kill(pid); err != nil {
-			return fmt.Errorf("failed to kill process: %v", err)
+	// Operate in terms of targetTime, so that we can pause in the middle of the operation
+	// to catch a new timeout (and possibly ignore that new timeout if it's not correct to
+	// take a new one).
+	targetTime := time.Now().Add(timeout)
+	killed := false
+	for !killed {
+		select {
+		case <-done:
+			return nil
+		case <-ctx.Done():
+			close(chControl)
+			return ctx.Err()
+		case <-time.After(time.Until(targetTime)):
+			close(chControl)
+			if ignoreKill {
+				return fmt.Errorf("timeout reached after %.0f seconds waiting for container process to exit",
+					timeout.Seconds())
+			}
+			pid, err := c.pid()
+			if err != nil {
+				return err
+			}
+			if err := Kill(pid); err != nil {
+				return fmt.Errorf("failed to kill process: %v", err)
+			}
+			killed = true
+		case newTimeout := <-c.stopTimeoutChan:
+			// If a new timeout comes in,
+			// interrupt the old one, and start a new one
+			newTargetTime := time.Now().Add(newTimeout)
+
+			// but only if it's earlier
+			if !newTargetTime.Before(targetTime) {
+				continue
+			}
+
+			targetTime = newTargetTime
+			timeout = newTimeout
```

Comment:
I get super twitchy when I see a select with no default, no matter how improbable it is to hit it.  At the very least I think you should have a debugf with a "we should never hit this" type of message.

Reply:
Unfortunately, I may need to leave you twitching, as it won't wait the timeout if there's a default
https://play.golang.org/p/hMqswDkjs8V